x/merkledb/proof.go

```diff
@@ -45,6 +53,68 @@ type ProofNode struct {
 	Children    map[byte]ids.ID
 }
 
+// Assumes [node.Key.KeyPath.NibbleLength] <= math.MaxUint64.
+func (node *ProofNode) ToProto() *syncpb.ProofNode {
+	pbNode := &syncpb.ProofNode{
+		Key: &syncpb.SerializedPath{
+			NibbleLength: uint64(node.KeyPath.NibbleLength),
+			Value:        node.KeyPath.Value,
+		},
+		Children: make(map[uint32][]byte, len(node.Children)),
+	}
+
+	for childIndex, childID := range node.Children {
+		childID := childID
+		pbNode.Children[uint32(childIndex)] = childID[:]
+	}
+
+	if node.ValueOrHash.hasValue {
+		pbNode.ValueOrHash = &syncpb.MaybeBytes{
+			Value: node.ValueOrHash.value,
+		}
+	} else {
+		pbNode.ValueOrHash = &syncpb.MaybeBytes{
+			IsNothing: true,
+		}
+	}
+
+	return pbNode
+}
+
+func (node *ProofNode) UnmarshalProto(pbNode *syncpb.ProofNode) error {
+	switch {
+	case pbNode == nil:
+		return ErrNilProofNode
+	case pbNode.ValueOrHash == nil:
+		return ErrNilValueOrHash
+	case pbNode.ValueOrHash.IsNothing && len(pbNode.ValueOrHash.Value) != 0:
+		return ErrInvalidMaybe
+	case pbNode.Key == nil:
+		return ErrNilSerializedPath
+	}
```

Comment:
For future reference: https://protobuf.dev/programming-guides/proto3/#default

> When a message is parsed, if the encoded message does not contain a particular singular element, the corresponding field in the parsed object is set to the default value for that field. These defaults are type-specific:
> 
> - ...
> - For message fields, the field is not set. Its exact value is language-dependent. See the [generated code guide](https://protobuf.dev/reference/) for details.

Which means we are enforcing that the message is specified - even if the inner fields are fully the zero value.

This is probably slightly inefficient w.r.t. network bandwidth. But I don't have strong opinions on it